dds/DCPS/DiscoveryBase.h

```diff
@@ -1198,7 +1198,7 @@ namespace OpenDDS {
       RcHandle<EndpointManagerSporadic> type_lookup_reply_deadline_processor_;
       TimeDuration max_type_lookup_service_reply_period_;
       DCPS::SequenceNumber type_lookup_service_sequence_number_;
-      bool use_xtypes_;
+      ACE_Atomic_Op<ACE_Thread_Mutex, bool> use_xtypes_;
```

Comment:
Whether or not it's "using" xtypes is determined in `Sedp::init` and shouldn't change after that.  Doesn't seem like it could be a thread synchronization problem.  If tsan says otherwise I'd be interested in seeing what it reported.

Reply:
I made use_xtypes_ const and then used a const_cast to write it in Sedp::init(), but ThreadSanitizer still didn't like it. Including the whole warning message below. Is it possible this leaves open a potential race between CPU caches, even when one thread creates / initialize the object if no synchronization mechanisms are used?

```
WARNING: ThreadSanitizer: data race (pid=11904)
  Write of size 1 at 0x7b6800001288 by main thread (mutexes: write M656, write M721):
    #0 OpenDDS::RTPS::Sedp::init(OpenDDS::DCPS::GUID_t const&, OpenDDS::RTPS::RtpsDiscovery const&, int, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Sedp.cpp:524:34 (libOpenDDS_Rtps.so+0x445724)
    #1 OpenDDS::RTPS::Spdp::init(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:228:10 (libOpenDDS_Rtps.so+0x4a09e7)
    #2 OpenDDS::RTPS::Spdp::Spdp(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:272:3 (libOpenDDS_Rtps.so+0x4a1821)
    #3 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::Spdp> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::Spdp, int, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t>, DDS::DomainParticipantQos, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> >(int const&, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t> const&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery* const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:291:26 (libOpenDDS_Rtps.so+0x41111e)
    #4 OpenDDS::RTPS::RtpsDiscovery::add_domain_participant(int, DDS::DomainParticipantQos const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:839:37 (libOpenDDS_Rtps.so+0x40d5ce)
    #5 OpenDDS::DCPS::DomainParticipantImpl::enable() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantImpl.cpp:1798:20 (libOpenDDS_Dcps.so+0x9ee3c9)
    #6 OpenDDS::DCPS::DomainParticipantFactoryImpl::create_participant(int, DDS::DomainParticipantQos const&, DDS::DomainParticipantListener*, unsigned int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantFactoryImpl.cpp:79:13 (libOpenDDS_Dcps.so+0x9db176)
    #7 main /storage/tsimpson/code/OpenDDS_pull/tests/DCPS/EntityLifecycleStress/subscriber.cpp:129:12 (subscriber+0x4bf722)

  Previous read of size 1 at 0x7b6800001288 by thread T2:
    #0 OpenDDS::RTPS::Sedp::DiscoveryReader::data_received_i(OpenDDS::DCPS::ReceivedDataSample const&, OpenDDS::DCPS::EntityId_t const&, OpenDDS::DCPS::Serializer&, OpenDDS::DCPS::Extensibility) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Sedp.cpp:4596:82 (libOpenDDS_Rtps.so+0x45b7b8)
    #1 OpenDDS::RTPS::Sedp::Reader::data_received(OpenDDS::DCPS::ReceivedDataSample const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Sedp.cpp:4505:5 (libOpenDDS_Rtps.so+0x45b304)
    #2 OpenDDS::DCPS::DataLink::data_received_i(OpenDDS::DCPS::ReceivedDataSample&, OpenDDS::DCPS::GUID_t const&, std::set<OpenDDS::DCPS::GUID_t, OpenDDS::DCPS::GUID_tKeyLessThan, std::allocator<OpenDDS::DCPS::GUID_t> > const&, OpenDDS::DCPS::ReceiveListenerSet::ConstrainReceiveSet) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/DataLink.cpp:711:17 (libOpenDDS_Dcps.so+0xb8a534)
    #3 OpenDDS::DCPS::DataLink::data_received(OpenDDS::DCPS::ReceivedDataSample&, OpenDDS::DCPS::GUID_t const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/DataLink.cpp:653:3 (libOpenDDS_Dcps.so+0xb8a19f)
    #4 OpenDDS::DCPS::RtpsUdpReceiveStrategy::deliver_sample_i(OpenDDS::DCPS::ReceivedDataSample&, OpenDDS::RTPS::Submessage const&, ACE_INET_Addr const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp:474:18 (libOpenDDS_Rtps_Udp.so+0x223278)
    #5 OpenDDS::DCPS::RtpsUdpReceiveStrategy::deliver_sample(OpenDDS::DCPS::ReceivedDataSample&, ACE_INET_Addr const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp:366:3 (libOpenDDS_Rtps_Udp.so+0x2226c3)
    #6 OpenDDS::DCPS::TransportReceiveStrategy<OpenDDS::DCPS::RtpsTransportHeader, OpenDDS::DCPS::RtpsSampleHeader>::handle_simple_dds_input(int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/TransportReceiveStrategy_T.cpp:217:11 (libOpenDDS_Rtps_Udp.so+0x226980)
    #7 OpenDDS::DCPS::RtpsUdpReceiveStrategy::handle_input(int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp:57:10 (libOpenDDS_Rtps_Udp.so+0x221340)
    #8 non-virtual thunk to OpenDDS::DCPS::RtpsUdpReceiveStrategy::handle_input(int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp (libOpenDDS_Rtps_Udp.so+0x221387)
    #9 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::notify_handle(int, unsigned long, ACE_Handle_Set&, ACE_Event_Handler*, int (ACE_Event_Handler::*)(int)) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:826:22 (libOpenDDS_Dcps.so+0xa78f58)
    #10 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::dispatch_io_set(int, int&, int, ACE_Handle_Set&, ACE_Handle_Set&, int (ACE_Event_Handler::*)(int)) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:1210:13 (libOpenDDS_Dcps.so+0xa78e19)
    #11 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::dispatch_io_handlers(ACE_Select_Reactor_Handle_Set&, int&, int&) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:1268:13 (libOpenDDS_Dcps.so+0xa78c7c)
    #12 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::dispatch(int, ACE_Select_Reactor_Handle_Set&) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:1372:22 (libOpenDDS_Dcps.so+0xa78935)
    #13 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::handle_events_i(ACE_Time_Value*) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:1463:15 (libOpenDDS_Dcps.so+0xa7ecf3)
    #14 ACE_Select_Reactor_T<ACE_Reactor_Token_T<ACE_Token> >::handle_events(ACE_Time_Value*) /storage/tsimpson/code/ACE_TAO/ACE/ace/Select_Reactor_T.cpp:1439:16 (libOpenDDS_Dcps.so+0xa75be2)
    #15 ACE_Reactor::run_reactor_event_loop(int (*)(ACE_Reactor*)) /storage/tsimpson/code/ACE_TAO/ACE/ace/Reactor.cpp:218:49 (libACE.so+0x1d840e)
    #16 OpenDDS::DCPS::ReactorTask::svc() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/ReactorTask.cpp:171:17 (libOpenDDS_Dcps.so+0xa72475)
    #17 virtual thunk to OpenDDS::DCPS::ReactorTask::svc() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/ReactorTask.cpp (libOpenDDS_Dcps.so+0xa72968)
    #18 ACE_Task_Base::svc_run(void*) /storage/tsimpson/code/ACE_TAO/ACE/ace/Task.cpp:269:29 (libACE.so+0x200aa5)
    #19 ACE_Thread_Adapter::invoke_i() /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Adapter.cpp:161:22 (libACE.so+0x201213)
    #20 ACE_Thread_Adapter::invoke() /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Adapter.cpp:96:16 (libACE.so+0x2010e7)
    #21 ace_thread_adapter /storage/tsimpson/code/ACE_TAO/ACE/ace/Base_Thread_Adapter.cpp:124:45 (libACE.so+0x13dfd2)

  Location is heap block of size 1408 at 0x7b6800001200 allocated by main thread:
    #0 operator new[](unsigned long, std::nothrow_t const&) <null> (subscriber+0x4bda72)
    #1 ACE_New_Allocator::malloc(unsigned long) /storage/tsimpson/code/ACE_TAO/ACE/ace/Malloc_Allocator.cpp:99:5 (libACE.so+0x18cbd5)
    #2 OpenDDS::DCPS::PoolAllocationBase::operator new(unsigned long) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/PoolAllocationBase.h:63:3 (subscriber+0x4c0ade)
    #3 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::Sedp> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::Sedp, OpenDDS::DCPS::GUID_t, OpenDDS::DCPS::reference_wrapper<OpenDDS::RTPS::Spdp>, OpenDDS::DCPS::reference_wrapper<ACE_Thread_Mutex> >(OpenDDS::DCPS::GUID_t const&, OpenDDS::DCPS::reference_wrapper<OpenDDS::RTPS::Spdp> const&, OpenDDS::DCPS::reference_wrapper<ACE_Thread_Mutex> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:279:22 (libOpenDDS_Rtps.so+0x4b76ad)
    #4 OpenDDS::RTPS::Spdp::Spdp(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:260:11 (libOpenDDS_Rtps.so+0x4a16c0)
    #5 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::Spdp> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::Spdp, int, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t>, DDS::DomainParticipantQos, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> >(int const&, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t> const&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery* const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:291:26 (libOpenDDS_Rtps.so+0x41111e)
    #6 OpenDDS::RTPS::RtpsDiscovery::add_domain_participant(int, DDS::DomainParticipantQos const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:839:37 (libOpenDDS_Rtps.so+0x40d5ce)
    #7 OpenDDS::DCPS::DomainParticipantImpl::enable() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantImpl.cpp:1798:20 (libOpenDDS_Dcps.so+0x9ee3c9)
    #8 OpenDDS::DCPS::DomainParticipantFactoryImpl::create_participant(int, DDS::DomainParticipantQos const&, DDS::DomainParticipantListener*, unsigned int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantFactoryImpl.cpp:79:13 (libOpenDDS_Dcps.so+0x9db176)
    #9 main /storage/tsimpson/code/OpenDDS_pull/tests/DCPS/EntityLifecycleStress/subscriber.cpp:129:12 (subscriber+0x4bf722)

  Mutex M656 (0x7b4400000038) created at:
    #0 pthread_mutex_init <null> (subscriber+0x4364a3)
    #1 ACE_OS::mutex_init(pthread_mutex_t*, int, char const*, pthread_mutexattr_t*, int, int) /storage/tsimpson/code/ACE_TAO/ACE/ace/OS_NS_Thread.cpp:1897:11 (libACE.so+0x1a8762)
    #2 ACE_OS::thread_mutex_init(pthread_mutex_t*, int, char const*, pthread_mutexattr_t*) /storage/tsimpson/code/ACE_TAO/ACE/ace/OS_NS_Thread.inl:3641:10 (libACE.so+0x1b1515)
    #3 ACE_Thread_Mutex::ACE_Thread_Mutex(char const*, pthread_mutexattr_t*) /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Mutex.cpp:49:7 (libACE.so+0x20a258)
    #4 OpenDDS::DCPS::PeerDiscovery<OpenDDS::RTPS::Spdp>::PeerDiscovery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DiscoveryBase.h:2472:16 (libOpenDDS_Rtps.so+0x40e949)
    #5 OpenDDS::RTPS::RtpsDiscovery::RtpsDiscovery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:100:5 (libOpenDDS_Rtps.so+0x409a24)
    #6 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::RtpsDiscovery> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::RtpsDiscovery, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:267:26 (libOpenDDS_Rtps.so+0x40ec38)
    #7 OpenDDS::RTPS::RtpsDiscovery::Config::discovery_config(ACE_Configuration_Heap&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:144:37 (libOpenDDS_Rtps.so+0x409e11)
    #8 OpenDDS::DCPS::Service_Participant::load_discovery_configuration(ACE_Configuration_Heap&, char const*) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/Service_Participant.cpp:2305:28 (libOpenDDS_Dcps.so+0xaa5fc1)
    #9 OpenDDS::DCPS::Service_Participant::load_configuration(ACE_Configuration_Heap&, char const*) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/Service_Participant.cpp:1512:18 (libOpenDDS_Dcps.so+0xaa65f4)
    #10 OpenDDS::DCPS::Service_Participant::load_configuration() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/Service_Participant.cpp:1467:20 (libOpenDDS_Dcps.so+0xaa2a53)
    #11 OpenDDS::DCPS::Service_Participant::get_domain_participant_factory(int&, char**) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/Service_Participant.cpp:422:21 (libOpenDDS_Dcps.so+0xaa1863)
    #12 main /storage/tsimpson/code/OpenDDS_pull/tests/DCPS/EntityLifecycleStress/subscriber.cpp:126:7 (subscriber+0x4bf6cf)

  Mutex M721 (0x7b6000000018) created at:
    #0 pthread_mutex_init <null> (subscriber+0x4364a3)
    #1 ACE_OS::mutex_init(pthread_mutex_t*, int, char const*, pthread_mutexattr_t*, int, int) /storage/tsimpson/code/ACE_TAO/ACE/ace/OS_NS_Thread.cpp:1897:11 (libACE.so+0x1a8762)
    #2 ACE_OS::thread_mutex_init(pthread_mutex_t*, int, char const*, pthread_mutexattr_t*) /storage/tsimpson/code/ACE_TAO/ACE/ace/OS_NS_Thread.inl:3641:10 (libACE.so+0x1b1515)
    #3 ACE_Thread_Mutex::ACE_Thread_Mutex(char const*, pthread_mutexattr_t*) /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Mutex.cpp:49:7 (libACE.so+0x20a258)
    #4 OpenDDS::DCPS::LocalParticipant<OpenDDS::RTPS::Sedp>::LocalParticipant(DDS::DomainParticipantQos const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DiscoveryBase.h:2043:7 (libOpenDDS_Rtps.so+0x4b74ae)
    #5 OpenDDS::RTPS::Spdp::Spdp(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:247:5 (libOpenDDS_Rtps.so+0x4a1505)
    #6 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::Spdp> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::Spdp, int, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t>, DDS::DomainParticipantQos, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> >(int const&, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t> const&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery* const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:291:26 (libOpenDDS_Rtps.so+0x41111e)
    #7 OpenDDS::RTPS::RtpsDiscovery::add_domain_participant(int, DDS::DomainParticipantQos const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:839:37 (libOpenDDS_Rtps.so+0x40d5ce)
    #8 OpenDDS::DCPS::DomainParticipantImpl::enable() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantImpl.cpp:1798:20 (libOpenDDS_Dcps.so+0x9ee3c9)
    #9 OpenDDS::DCPS::DomainParticipantFactoryImpl::create_participant(int, DDS::DomainParticipantQos const&, DDS::DomainParticipantListener*, unsigned int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantFactoryImpl.cpp:79:13 (libOpenDDS_Dcps.so+0x9db176)
    #10 main /storage/tsimpson/code/OpenDDS_pull/tests/DCPS/EntityLifecycleStress/subscriber.cpp:129:12 (subscriber+0x4bf722)

  Thread T2 (tid=11973, running) created by main thread at:
    #0 pthread_create <null> (subscriber+0x435395)
    #1 ACE_OS::thr_create(void* (*)(void*), void*, long, unsigned long*, unsigned long*, long, void*, unsigned long, ACE_Base_Thread_Adapter*, char const**) /storage/tsimpson/code/ACE_TAO/ACE/ace/OS_NS_Thread.cpp:3936:3 (libACE.so+0x1a9e85)
    #2 ACE_Thread::spawn(void* (*)(void*), void*, long, unsigned long*, unsigned long*, long, void*, unsigned long, ACE_Thread_Adapter*, char const**) /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread.inl:87:10 (libACE.so+0x207a25)
    #3 ACE_Thread_Manager::spawn_i(void* (*)(void*), void*, long, unsigned long*, unsigned long*, long, int, void*, unsigned long, ACE_Task_Base*, char const**) /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Manager.cpp:640:22 (libACE.so+0x203727)
    #4 ACE_Thread_Manager::spawn_n(unsigned long, void* (*)(void*), void*, long, long, int, ACE_Task_Base*, unsigned long*, void**, unsigned long*, char const**) /storage/tsimpson/code/ACE_TAO/ACE/ace/Thread_Manager.cpp:758:17 (libACE.so+0x203c29)
    #5 ACE_Task_Base::activate(long, int, int, long, int, ACE_Task_Base*, unsigned long*, void**, unsigned long*, unsigned long*, char const**) /storage/tsimpson/code/ACE_TAO/ACE/ace/Task.cpp:159:23 (libACE.so+0x2009b8)
    #6 OpenDDS::DCPS::ReactorTask::open_reactor_task(void*, OpenDDS::DCPS::TimeDuration, OpenDDS::DCPS::ThreadStatusManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/ReactorTask.cpp:106:7 (libOpenDDS_Dcps.so+0xa71fb6)
    #7 OpenDDS::DCPS::TransportImpl::create_reactor_task(bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/TransportImpl.cpp:116:22 (libOpenDDS_Dcps.so+0xbccd86)
    #8 OpenDDS::DCPS::RtpsUdpTransport::configure_i(OpenDDS::DCPS::RtpsUdpInst&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpTransport.cpp:528:3 (libOpenDDS_Rtps_Udp.so+0x23433a)
    #9 OpenDDS::DCPS::RtpsUdpTransport::RtpsUdpTransport(OpenDDS::DCPS::RtpsUdpInst&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpTransport.cpp:44:10 (libOpenDDS_Rtps_Udp.so+0x233dde)
    #10 OpenDDS::DCPS::RcHandle<OpenDDS::DCPS::RtpsUdpTransport> OpenDDS::DCPS::make_rch<OpenDDS::DCPS::RtpsUdpTransport, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::RtpsUdpInst> >(OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::RtpsUdpInst> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:267:26 (libOpenDDS_Rtps_Udp.so+0x21d488)
    #11 OpenDDS::DCPS::RtpsUdpInst::new_impl() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/rtps_udp/RtpsUdpInst.cpp:61:10 (libOpenDDS_Rtps_Udp.so+0x219522)
    #12 OpenDDS::DCPS::TransportInst::impl() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/TransportInst.cpp:126:15 (libOpenDDS_Dcps.so+0xbcfc95)
    #13 OpenDDS::DCPS::TransportInst::reactor_task() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/transport/framework/TransportInst.cpp:203:49 (libOpenDDS_Dcps.so+0xbd0125)
    #14 OpenDDS::RTPS::Sedp::init(OpenDDS::DCPS::GUID_t const&, OpenDDS::RTPS::RtpsDiscovery const&, int, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Sedp.cpp:411:36 (libOpenDDS_Rtps.so+0x444f34)
    #15 OpenDDS::RTPS::Spdp::init(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:228:10 (libOpenDDS_Rtps.so+0x4a09e7)
    #16 OpenDDS::RTPS::Spdp::Spdp(int, OpenDDS::DCPS::GUID_t&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Spdp.cpp:272:3 (libOpenDDS_Rtps.so+0x4a1821)
    #17 OpenDDS::DCPS::RcHandle<OpenDDS::RTPS::Spdp> OpenDDS::DCPS::make_rch<OpenDDS::RTPS::Spdp, int, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t>, DDS::DomainParticipantQos, OpenDDS::RTPS::RtpsDiscovery*, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> >(int const&, OpenDDS::DCPS::reference_wrapper<OpenDDS::DCPS::GUID_t> const&, DDS::DomainParticipantQos const&, OpenDDS::RTPS::RtpsDiscovery* const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService> const&) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RcHandle_T.h:291:26 (libOpenDDS_Rtps.so+0x41111e)
    #18 OpenDDS::RTPS::RtpsDiscovery::add_domain_participant(int, DDS::DomainParticipantQos const&, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/RtpsDiscovery.cpp:839:37 (libOpenDDS_Rtps.so+0x40d5ce)
    #19 OpenDDS::DCPS::DomainParticipantImpl::enable() /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantImpl.cpp:1798:20 (libOpenDDS_Dcps.so+0x9ee3c9)
    #20 OpenDDS::DCPS::DomainParticipantFactoryImpl::create_participant(int, DDS::DomainParticipantQos const&, DDS::DomainParticipantListener*, unsigned int) /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/DomainParticipantFactoryImpl.cpp:79:13 (libOpenDDS_Dcps.so+0x9db176)
    #21 main /storage/tsimpson/code/OpenDDS_pull/tests/DCPS/EntityLifecycleStress/subscriber.cpp:129:12 (subscriber+0x4bf722)

SUMMARY: ThreadSanitizer: data race /storage/tsimpson/code/OpenDDS_pull/dds/DCPS/RTPS/Sedp.cpp:524:34 in OpenDDS::RTPS::Sedp::init(OpenDDS::DCPS::GUID_t const&, OpenDDS::RTPS::RtpsDiscovery const&, int, OpenDDS::DCPS::RcHandle<OpenDDS::XTypes::TypeLookupService>)
```

Reply:
I was able to resolve this data race by moving the initialization of use_xtypes_ to before the reactor task / reactor thread creation.